File: crates/blazen-node/examples/llm_openai.ts
/**
 * OpenAI-powered content pipeline with Blazen.
 *
 * Demonstrates using Blazen's CompletionModel with OpenAI to build a real
 * 3-step content generation workflow:
 *
 *   1. generate_outline  -- GPT creates a blog post outline from a topic
 *   2. write_draft       -- GPT writes a short draft based on the outline
 *   3. review            -- GPT provides editorial feedback on the draft
 *
 * The example also tracks cumulative token usage across all LLM calls using
 * Blazen's context API (ctx.set / ctx.get).
 *
 * Run with: OPENAI_API_KEY=sk-... npx tsx llm_openai.ts
 */

import { Workflow, CompletionModel } from "blazen";
import type { Context, JsWorkflowResult } from "blazen";

// ---------------------------------------------------------------------------
// Usage tracking interface
// ---------------------------------------------------------------------------
interface UsageInfo {
  promptTokens: number;
  completionTokens: number;
  totalTokens: number;
}

// ---------------------------------------------------------------------------
// Helper: accumulate token usage in context
// ---------------------------------------------------------------------------
async function trackUsage(ctx: Context, usage: UsageInfo): Promise<void> {
  const prev: UsageInfo = (await ctx.get("total_usage")) || {
    promptTokens: 0,
    completionTokens: 0,
    totalTokens: 0,
  };
  await ctx.set("total_usage", {
    promptTokens: prev.promptTokens + (usage.promptTokens || 0),
    completionTokens: prev.completionTokens + (usage.completionTokens || 0),
    totalTokens: prev.totalTokens + (usage.totalTokens || 0),
  });
}

// ---------------------------------------------------------------------------
// Validate that the API key is available.
// ---------------------------------------------------------------------------
const apiKey: string | undefined = process.env.OPENAI_API_KEY;
if (!apiKey) {
  console.error(
    "ERROR: OPENAI_API_KEY environment variable is not set.\n" +
      "Run with: OPENAI_API_KEY=sk-... npx tsx llm_openai.ts"
  );
  process.exit(1);
}

// Create the OpenAI completion model (gpt-5.3-chat-latest is cheap for examples).
const model: CompletionModel = CompletionModel.openai(apiKey);

// Build the 3-step content pipeline.
const wf: Workflow = new Workflow("content-pipeline");

// ---------------------------------------------------------------------------
// Step 1: Generate an outline
// ---------------------------------------------------------------------------
wf.addStep(
  "generate_outline",
  ["blazen::StartEvent"],
  async (event: Record<string, any>, ctx: Context) => {
    const topic: string = event.topic;

    // Store the model reference key in shared context so downstream steps
    // can retrieve it. We store the topic too for convenience.
    await ctx.set("topic", topic);

    console.log(`[generate_outline] Requesting outline for: "${topic}"`);

    const response: Record<string, any> = await model.completeWithOptions(
      [
        {
          role: "system",
          content:
            "You are a professional content strategist. " +
            "When given a topic, produce a concise blog post outline " +
            "with 3-5 sections. Keep it brief -- just section titles " +
            "and one-line descriptions.",
        },
        {
          role: "user",
          content: `Create a blog post outline about: ${topic}`,
        },
      ],
      { temperature: 0.7, maxTokens: 512, model: "gpt-5.3-chat-latest" }
    );

    const outline: string = response.content;
    const usage: UsageInfo = response.usage;
    await trackUsage(ctx, usage);

    console.log(
      `[generate_outline] Outline received (${usage.totalTokens} tokens)`
    );
    console.log(`\n--- Outline ---\n${outline}\n`);

    return {
      type: "OutlineEvent",
      outline: outline,
      topic: topic,
    };
  }
);

// ---------------------------------------------------------------------------
// Step 2: Write a short draft
// ---------------------------------------------------------------------------
wf.addStep(
  "write_draft",
  ["OutlineEvent"],
  async (event: Record<string, any>, ctx: Context) => {
    console.log("[write_draft] Writing draft from outline...");

    const response: Record<string, any> = await model.completeWithOptions(
      [
        {
          role: "system",
          content:
            "You are a skilled blog writer. Given an outline, write a " +
            "short but engaging first draft (2-3 paragraphs). " +
            "Keep the tone approachable and informative.",
        },
        {
          role: "user",
          content:
            `Topic: ${event.topic}\n\nOutline:\n${event.outline}\n\n` +
            "Write a short draft based on this outline.",
        },
      ],
      { temperature: 0.8, maxTokens: 768, model: "gpt-5.3-chat-latest" }
    );

    const draft: string = response.content;
    const usage: UsageInfo = response.usage;
    await trackUsage(ctx, usage);

    console.log(
      `[write_draft] Draft received (${usage.totalTokens} tokens)`
    );
    console.log(`\n--- Draft ---\n${draft}\n`);

    return {
      type: "DraftEvent",
      draft: draft,
      topic: event.topic,
    };
  }
);

// ---------------------------------------------------------------------------
// Step 3: Editorial review
// ---------------------------------------------------------------------------
wf.addStep(
  "review",
  ["DraftEvent"],
  async (event: Record<string, any>, ctx: Context) => {
    console.log("[review] Reviewing draft...");

    const response: Record<string, any> = await model.completeWithOptions(
      [
        {
          role: "system",
          content:
            "You are a senior editor. Review the following blog draft " +
            "and provide 3-5 bullet points of constructive feedback. " +
            "Be specific and actionable.",
        },
        {
          role: "user",
          content:
            `Topic: ${event.topic}\n\nDraft:\n${event.draft}\n\n` +
            "Provide editorial feedback.",
        },
      ],
      { temperature: 0.5, maxTokens: 512, model: "gpt-5.3-chat-latest" }
    );

    const feedback: string = response.content;
    const usage: UsageInfo = response.usage;
    await trackUsage(ctx, usage);

    console.log(
      `[review] Feedback received (${usage.totalTokens} tokens)`
    );
    console.log(`\n--- Editorial Feedback ---\n${feedback}\n`);

    const totalUsage: UsageInfo = await ctx.get("total_usage");

    return {
      type: "blazen::StopEvent",
      result: {
        topic: event.topic,
        draft: event.draft,
        feedback: feedback,
        totalUsage: totalUsage,
      },
    };
  }
);

// ---------------------------------------------------------------------------
// Run the workflow
// ---------------------------------------------------------------------------
const topic: string =
  "Why Rust-powered Python libraries are the future of performance";

const result: JsWorkflowResult = await wf.run({ topic: topic });

// Print summary.
console.log("=".repeat(60));
console.log("CONTENT PIPELINE COMPLETE");
console.log("=".repeat(60));

const outputUsage: UsageInfo = result.data.totalUsage;
console.log(`Topic:              ${result.data.topic}`);
console.log(`Prompt tokens:      ${outputUsage.promptTokens}`);
console.log(`Completion tokens:  ${outputUsage.completionTokens}`);
console.log(`Total tokens:       ${outputUsage.totalTokens}`);

File: crates/blazen-py/examples/llm_openai.py
"""OpenAI-powered content pipeline with Blazen.

Demonstrates using Blazen's CompletionModel with OpenAI to build a real
3-step content generation workflow:

  1. generate_outline  -- GPT creates a blog post outline from a topic
  2. write_draft       -- GPT writes a short draft based on the outline
  3. review            -- GPT provides editorial feedback on the draft

The example also tracks cumulative token usage across all LLM calls using
Blazen's synchronous context API (ctx.set / ctx.get).

Run with: OPENAI_API_KEY=sk-... python llm_openai.py
"""

import asyncio
import os
import sys

from blazen import (
    ChatMessage,
    CompletionModel,
    Context,
    Event,
    ProviderOptions,
    StartEvent,
    StopEvent,
    Workflow,
    step,
)


# ---------------------------------------------------------------------------
# Helper: accumulate token usage in context
# ---------------------------------------------------------------------------
def track_usage(ctx: Context, usage: dict) -> None:
    """Add a single LLM response's token counts to the running totals."""
    prev = ctx.get("total_usage") or {
        "prompt_tokens": 0,
        "completion_tokens": 0,
        "total_tokens": 0,
    }
    ctx.set("total_usage", {
        "prompt_tokens": prev["prompt_tokens"] + (usage.get("prompt_tokens") or 0),
        "completion_tokens": prev["completion_tokens"] + (usage.get("completion_tokens") or 0),
        "total_tokens": prev["total_tokens"] + (usage.get("total_tokens") or 0),
    })


# ---------------------------------------------------------------------------
# Step 1: Generate an outline
# ---------------------------------------------------------------------------
# Module-level model reference. CompletionModel is a native PyO3 object and
# cannot be serialised to JSON, so we store it here instead of in context.
MODEL: CompletionModel | None = None


@step
async def generate_outline(ctx: Context, ev: Event) -> Event:
    """Ask GPT to produce a concise blog post outline for the given topic."""
    topic: str = ev.topic

    print(f"[generate_outline] Requesting outline for: {topic!r}")

    response = await MODEL.complete(
        [
            ChatMessage.system(
                "You are a professional content strategist. "
                "When given a topic, produce a concise blog post outline "
                "with 3-5 sections. Keep it brief -- just section titles "
                "and one-line descriptions."
            ),
            ChatMessage.user(f"Create a blog post outline about: {topic}"),
        ],
        temperature=0.7,
        max_tokens=512,
    )

    outline = response["content"]
    track_usage(ctx, response["usage"])

    print(f"[generate_outline] Outline received ({response['usage']['total_tokens']} tokens)")
    print(f"\n--- Outline ---\n{outline}\n")

    return Event("OutlineEvent", outline=outline, topic=topic)


# ---------------------------------------------------------------------------
# Step 2: Write a short draft
# ---------------------------------------------------------------------------
@step(accepts=["OutlineEvent"])
async def write_draft(ctx: Context, ev: Event) -> Event:
    """Ask GPT to write a short first draft based on the outline."""

    print("[write_draft] Writing draft from outline...")

    response = await MODEL.complete(
        [
            ChatMessage.system(
                "You are a skilled blog writer. Given an outline, write a "
                "short but engaging first draft (2-3 paragraphs). "
                "Keep the tone approachable and informative."
            ),
            ChatMessage.user(
                f"Topic: {ev.topic}\n\nOutline:\n{ev.outline}\n\n"
                "Write a short draft based on this outline."
            ),
        ],
        temperature=0.8,
        max_tokens=768,
    )

    draft = response["content"]
    track_usage(ctx, response["usage"])

    print(f"[write_draft] Draft received ({response['usage']['total_tokens']} tokens)")
    print(f"\n--- Draft ---\n{draft}\n")

    return Event("DraftEvent", draft=draft, topic=ev.topic)


# ---------------------------------------------------------------------------
# Step 3: Editorial review
# ---------------------------------------------------------------------------
@step(accepts=["DraftEvent"])
async def review(ctx: Context, ev: Event) -> StopEvent:
    """Ask GPT to provide brief editorial feedback on the draft."""
    print("[review] Reviewing draft...")

    response = await MODEL.complete(
        [
            ChatMessage.system(
                "You are a senior editor. Review the following blog draft "
                "and provide 3-5 bullet points of constructive feedback. "
                "Be specific and actionable."
            ),
            ChatMessage.user(
                f"Topic: {ev.topic}\n\nDraft:\n{ev.draft}\n\n"
                "Provide editorial feedback."
            ),
        ],
        temperature=0.5,
        max_tokens=512,
    )

    feedback = response["content"]
    track_usage(ctx, response["usage"])

    print(f"[review] Feedback received ({response['usage']['total_tokens']} tokens)")
    print(f"\n--- Editorial Feedback ---\n{feedback}\n")

    total_usage = ctx.get("total_usage")

    return StopEvent(result={
        "topic": ev.topic,
        "draft": ev.draft,
        "feedback": feedback,
        "total_usage": total_usage,
    })


# ---------------------------------------------------------------------------
# Main entry point
# ---------------------------------------------------------------------------
async def main() -> None:
    # Validate that the API key is available.
    api_key = os.environ.get("OPENAI_API_KEY")
    if not api_key:
        print(
            "ERROR: OPENAI_API_KEY environment variable is not set.\n"
            "Run with: OPENAI_API_KEY=sk-... python llm_openai.py",
            file=sys.stderr,
        )
        sys.exit(1)

    # Create the OpenAI completion model (gpt-5.3-chat-latest is cheap for examples).
    # Stored at module level because CompletionModel is a native object that
    # cannot be serialised into the JSON-based workflow context.
    global MODEL
    MODEL = CompletionModel.openai(options=ProviderOptions(api_key=api_key, model="gpt-5.3-chat-latest"))

    # Build the 3-step content pipeline.
    wf = Workflow("content-pipeline", [generate_outline, write_draft, review])

    topic = "Why Rust-powered Python libraries are the future of performance"
    handler = await wf.run(topic=topic)
    result = await handler.result()
    output = result.to_dict()

    # Print summary.
    print("=" * 60)
    print("CONTENT PIPELINE COMPLETE")
    print("=" * 60)

    usage = output["result"]["total_usage"]
    print(f"Topic:              {output['result']['topic']}")
    print(f"Prompt tokens:      {usage['prompt_tokens']}")
    print(f"Completion tokens:  {usage['completion_tokens']}")
    print(f"Total tokens:       {usage['total_tokens']}")


if __name__ == "__main__":
    asyncio.run(main())
